storage/Devices/BlkDeviceImpl.cc

```diff
@@ -478,17 +478,45 @@ namespace storage
     Encryption*
     BlkDevice::Impl::create_encryption(const string& dm_name)
     {
-	if (num_children() != 0)
-	    ST_THROW(WrongNumberOfChildren(num_children(), 0));
+	Devicegraph* devicegraph = get_devicegraph();
+
+	vector<Devicegraph::Impl::edge_descriptor> out_edges =
+	    devicegraph->get_impl().out_edges(get_vertex());
+
+	Encryption* encryption = Luks::create(devicegraph, dm_name);
+	Devicegraph::Impl::vertex_descriptor encryption_vertex = encryption->get_impl().get_vertex();
 
-	Luks* luks = Luks::create(get_devicegraph(), dm_name);
+	User::create(devicegraph, get_non_impl(), encryption);
 
-	User::create(get_devicegraph(), get_non_impl(), luks);
+	for (Devicegraph::Impl::edge_descriptor out_edge : out_edges)
+	{
+	    devicegraph->get_impl().set_source(out_edge, encryption_vertex);
+	}
 
 	// TODO maybe add parent_added() next to parent_has_new_region() for this?
-	luks->get_impl().parent_has_new_region(get_non_impl());
+	encryption->get_impl().parent_has_new_region(get_non_impl());
+
+	return encryption;
+    }
+
+
+    void
+    BlkDevice::Impl::remove_encryption()
+    {
+	Devicegraph* devicegraph = get_devicegraph();
+
+	Encryption* encryption = get_encryption();
```

Comment:
I assume if there is no encryption layer on this level, this will throw an exception, right?
In that case: Do we have a corresponding has_encryption() predicate?

Reply:
Disregard. I just found has_encryption() below.

Reply:
Yes, we have has_encryption() and get_encryption() will throw.